### PR TITLE
Move SDK system APIs to the new runtime types

### PR DESCRIPTION
**Pull request by jvff**
## Motivation

Interfacing with the host from guest Wasm applications should follow a unique and discoverable interface. New runtime types were introduced for this purpose (#1748), and the `system_api` modules in the SDK should now be removed, with its contents moved to the new types.

## Proposal

Remove the `system_api` modules from `linera-sdk`, moving all the system APIs to the `ContractRuntime` and the `ServiceRuntime` types. Also remove the default functions from the `Contract` and `Service` traits that interface with the host under the hood.

## Test Plan

This refactors the system APIs, and existing test should catch regressions.

## Release Plan

Includes major changes to the SDK, so
- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: examples/amm/src/contract.rs [hunk 1]
```diff
@@ -346,6 +352,7 @@ impl Amm {
 
     fn execute_application_call_remote(
         &mut self,
+        runtime: &mut ContractRuntime<Abi>,
         outcome: &mut ExecutionOutcome<Message>,
```
Comment (ma2bd) [inline examples/amm/src/contract.rs:356]: We kinda have the worst of both worlds right after this PR :-|

**Review by ma2bd — APPROVED** [review]

File: examples/amm/src/contract.rs [hunk 2]
```diff
@@ -6,12 +6,11 @@
 mod state;
 
 use self::state::Amm;
-use amm::{AmmError, ApplicationCall, Message, Operation};
+use amm::{AmmAbi as Abi, AmmError, ApplicationCall, Message, Operation};
```
Comment (MathieuDutSik) [inline examples/amm/src/contract.rs:9]: Do we really need that `X as Y`.

File: examples/amm/src/contract.rs [hunk 3]
```diff
@@ -31,33 +30,33 @@ impl Contract for Amm {
...
+            self.execute_order_remote(runtime, &mut outcome, operation)?;
         }
 
         Ok(outcome)
     }
 
     async fn execute_message(
         &mut self,
-        runtime: &mut ContractRuntime,
+        runtime: &mut ContractRuntime<Abi>,
```
Comment (MathieuDutSik) [inline examples/amm/src/contract.rs:59]: It is not totally great to have a `runtime` being put all over. Could the runtime be part of the self? Not in this PR of course.

File: examples/amm/src/contract.rs (hunk 1 above)
Comment (jvff) [reply examples/amm/src/contract.rs]: Updated. Hopefully it's better now :crossed_fingers:

File: examples/amm/src/contract.rs (hunk 2 above)
Comment (jvff) [reply examples/amm/src/contract.rs]: Removed :tada:

File: examples/amm/src/contract.rs (hunk 3 above)
Comment (jvff) [reply examples/amm/src/contract.rs]: Done, thanks!

File: linera-sdk/wasm-tests/src/lib.rs [hunk 4]
```diff
@@ -52,12 +56,12 @@ fn mock_application_id() {
...
-        contract::system_api::current_application_id(),
-        application_id
+        test_contract_runtime::<TestApp>().application_id(),
+        application_id.with_abi()
     );
     assert_eq!(
-        service::system_api::current_application_id(),
-        application_id
+        test_service_runtime::<TestApp>().application_id(),
+        application_id.with_abi()
```
Comment (afck) [inline linera-sdk/wasm-tests/src/lib.rs:64]: :heart:

File: linera-sdk/wasm-tests/src/lib.rs [hunk 5]
```diff
@@ -68,16 +72,13 @@ fn mock_application_parameters() {
...
-        contract::system_api::private::current_application_parameters(),
-        serialized_parameters
+        test_contract_runtime::<TestApp>().application_parameters(),
+        parameters
     );
     assert_eq!(
-        service::system_api::private::current_application_parameters(),
-        serialized_parameters
+        test_service_runtime::<TestApp>().application_parameters(),
+        parameters
```
Comment (afck) [inline linera-sdk/wasm-tests/src/lib.rs:81]: :star_struck:

File: examples/meta-counter/src/service.rs [hunk 6]
```diff
@@ -25,13 +26,16 @@ impl Service for MetaCounterService {
     type Storage = SimpleStateStorage<Self>;
     type State = MetaCounter;
 
-    async fn new(state: Self::State, _runtime: ServiceRuntime) -> Result<Self, Self::Error> {
-        Ok(MetaCounterService { _state: state })
+    async fn new(state: Self::State, runtime: ServiceRuntime<Self>) -> Result<Self, Self::Error> {
+        Ok(MetaCounterService {
+            _state: state,
```
Comment (afck) [inline examples/meta-counter/src/service.rs:31]: Do we even need that field, or should we drop the state view right away?

File: examples/matching-engine/src/contract.rs [hunk 7]
```diff
@@ -211,20 +215,20 @@ impl MatchingEngineContract {
         amount: &Amount,
         nature: &OrderNature,
         price: &Price,
-    ) -> Result<(), MatchingEngineError> {
+    ) {
```
Comment (afck) [inline examples/matching-engine/src/contract.rs:218]: (Good catch!)

File: examples/counter/src/contract.rs [hunk 8]
```diff
@@ -41,9 +42,6 @@ impl Contract for CounterContract {
         &mut self,
         value: u64,
     ) -> Result<ExecutionOutcome<Self::Message>, Self::Error> {
-        // Validate that the application parameters were configured correctly.
-        assert!(Self::parameters().is_ok());
```
Comment (afck) [inline examples/counter/src/contract.rs:45]: Should we call `let _ = self.runtime.application_parameters();`, like in crowd-funding?

File: linera-sdk/src/contract/runtime.rs [hunk 9]
```diff
@@ -86,4 +110,64 @@ impl ContractRuntime {
...
+            .timestamp
+            .get_or_insert_with(|| wit::read_system_timestamp().into())
+    }
+
+    /// Returns the current chain balance.
+    pub fn chain_balance(&mut self) -> Amount {
+        wit::read_chain_balance().into()
+    }
+
+    /// Returns the balance of one of the chain owners.
```
Comment (afck) [inline linera-sdk/src/contract/runtime.rs:126]: I think also people who aren't chain owners can have a balance?

**Review by afck — APPROVED** [review]
I love the added type safety! :tada:

File: examples/meta-counter/src/service.rs (hunk 6 above)
Comment (jvff) [reply examples/meta-counter/src/service.rs]: Good point, thanks!

File: linera-sdk/src/contract/runtime.rs (hunk 9 above)
Comment (jvff) [reply linera-sdk/src/contract/runtime.rs]: Oops, my bad. I fixed the documentation.

File: linera-sdk/src/service/runtime.rs [hunk 10]
```diff
@@ -33,6 +70,57 @@ impl ServiceRuntime {
...
+    pub fn system_time(&self) -> Timestamp {
+        Self::fetch_value_through_cache(&self.timestamp, || wit::read_system_timestamp().into())
+    }
+
+    /// Returns the current chain balance.
+    pub fn chain_balance(&self) -> Amount {
+        Self::fetch_value_through_cache(&self.chain_balance, || wit::read_chain_balance().into())
+    }
+
+    /// Returns the balance of one of the chain owners.
```
Comment (afck) [inline linera-sdk/src/service/runtime.rs:83]: But they aren't necessarily a chain owner!

**Review by afck — APPROVED** [review]

File: linera-sdk/src/service/runtime.rs (hunk 10 above)
Comment (jvff) [reply linera-sdk/src/service/runtime.rs]: Oops! Fixed in `ContractRuntime` but forgot to fix here as well. Now it's done :sweat_smile:

**Pull request merged**